File: python/aihash/store.py
"""Раскладка на диске и кодирование записей.

Обычные файлы, без базы данных. Требование из плана: через семь лет
посторонний эксперт открывает каталог и понимает содержимое без нашей
инфраструктуры.
"""

from __future__ import annotations

import base64
import json
import os
from typing import Dict, List, Optional

from . import core

RECORD_V = 1


# --- кодирование значения поля ---------------------------------------------


def encode_field(name: str, salt: bytes, value) -> dict:
    out = {"name": name, "salt": salt.hex()}
    if isinstance(value, str):
        out["s"] = value
    elif isinstance(value, (bytes, bytearray)):
        out["b"] = base64.b64encode(bytes(value)).decode("ascii")
    else:
        raise core.FormatError(
            "значение поля %r должно быть str или bytes, получено %s"
            % (name, type(value).__name__))
    return out


def raw_value(f: dict) -> bytes:
    if "s" in f and "b" in f:
        raise core.FormatError("поле %r: s и b одновременно" % f.get("name"))
    if "s" in f:
        return f["s"].encode("utf-8")
    if "b" in f:
        return base64.b64decode(f["b"])
    raise core.FormatError("поле %r: нет ни s, ни b" % f.get("name"))


def fields_for_root(fields: List[dict]) -> List[dict]:
    """Приводит поля из файла к виду, который принимает core.content_root."""
    out = []
    for f in fields:
        if "leaf" in f:
            if "salt" in f or "s" in f or "b" in f:
                raise core.FormatError(
                    "вычеркнутое поле %r сохранило соль или значение" % f.get("name"))
            out.append({"name": f["name"], "leaf": bytes.fromhex(f["leaf"])})
        else:
            out.append({"name": f["name"], "salt": bytes.fromhex(f["salt"]),
                        "value": raw_value(f)})
    return out


def record_line(seq: int, ts: int, fields: List[dict], croot: bytes,
                link: bytes) -> str:
    return json.dumps({"v": RECORD_V, "seq": seq, "ts": ts, "fields": fields,
                       "content_root": croot.hex(), "link": link.hex()},
                      ensure_ascii=False, separators=(",", ":"))


# --- атомарная запись файла -------------------------------------------------


def write_json(path: str, obj: dict) -> None:
    tmp = path + ".tmp"
    with open(tmp, "w", encoding="utf-8") as f:
        json.dump(obj, f, ensure_ascii=False, indent=2)
        f.write("\n")
        f.flush()
        os.fsync(f.fileno())
    os.replace(tmp, path)


def read_json(path: str) -> dict:
    with open(path, encoding="utf-8") as f:
        return json.load(f)


# --- раскладка --------------------------------------------------------------


class Layout:
    def __init__(self, root: str):
        self.root = os.path.abspath(root)

    @property
    def realm_file(self) -> str:
        return os.path.join(self.root, "realm.json")

    def stream_dir(self, stream_id: str) -> str:
        return os.path.join(self.root, "streams", stream_id)

    def meta_file(self, stream_id: str) -> str:
        return os.path.join(self.stream_dir(stream_id), "meta.json")

    def segments_dir(self, stream_id: str) -> str:
        return os.path.join(self.stream_dir(stream_id), "segments")

    def segment_file(self, stream_id: str, period: str) -> str:
        return os.path.join(self.segments_dir(stream_id), period + ".jsonl")

    def segment_ckpt_file(self, stream_id: str, period: str) -> str:
        return os.path.join(self.segments_dir(stream_id), period + ".seg.json")

    def day_file(self, date: str) -> str:
        return os.path.join(self.root, "days", date + ".day.json")

    def anchor_file(self, date: str, kind: str, ext: str) -> str:
        return os.path.join(self.root, "anchors", "%s.%s.%s" % (date, kind, ext))

    def anchors_dir(self) -> str:
        return os.path.join(self.root, "anchors")

    # --- перечисление ---

    def streams(self) -> List[str]:
        d = os.path.join(self.root, "streams")
        if not os.path.isdir(d):
            return []
        return sorted(n for n in os.listdir(d) if os.path.isdir(os.path.join(d, n)))

    def periods(self, stream_id: str) -> List[str]:
        d = self.segments_dir(stream_id)
        if not os.path.isdir(d):
            return []
        return sorted(n[:-6] for n in os.listdir(d) if n.endswith(".jsonl"))

    def dates(self) -> List[str]:
        d = os.path.join(self.root, "days")
        if not os.path.isdir(d):
            return []
        return sorted(n[:-9] for n in os.listdir(d) if n.endswith(".day.json"))

    def anchors_for(self, date: str) -> List[str]:
        d = self.anchors_dir()
        if not os.path.isdir(d):
            return []
        return sorted(os.path.join(d, n) for n in os.listdir(d)
                      if n.startswith(date + "."))

    # --- чтение ---

    def read_segment(self, stream_id: str, period: str) -> List[dict]:
        path = self.segment_file(stream_id, period)
        if not os.path.exists(path):
            return []
        out = []
        with open(path, encoding="utf-8") as f:
            for lineno, line in enumerate(f, 1):
                line = line.strip()
                if not line:
                    continue
                try:
                    out.append(json.loads(line))
                except json.JSONDecodeError as e:
                    raise core.FormatError(
                        "%s строка %d: не разбирается как JSON (%s)"
                        % (path, lineno, e))
        return out

    def read_segment_tail(self, stream_id: str, period: str) -> Optional[dict]:
        """Последняя запись отрезка без чтения всего файла.

        Пишущему нужен только кончик цепи. Читать ради него весь журнал —
        значит превращать запуск процесса в минуты на журнале за несколько лет.
        Проверка цепи целиком — работа верификатора, а не писателя.
        """
        path = self.segment_file(stream_id, period)
        if not os.path.exists(path):
            return None
        size = os.path.getsize(path)
        if size == 0:
            return None
        window = 1 << 16
        while True:
            with open(path, "rb") as f:
                f.seek(max(0, size - window))
                chunk = f.read()
            if not chunk.endswith(b"\n"):
                raise core.FormatError(
                    "%s: последняя строка оборвана — процесс упал посреди записи. "
                    "Обрежьте файл до последнего перевода строки; всё до него "
                    "проверяется как обычно." % path)
            lines = [l for l in chunk.split(b"\n") if l.strip()]
            if lines:
                try:
                    return json.loads(lines[-1].decode("utf-8"))
                except (UnicodeDecodeError, json.JSONDecodeError):
                    pass
            if window >= size:
                raise core.FormatError("%s: последняя запись не разбирается" % path)
            window = min(window * 8, size)

    def read_segment_ckpt(self, stream_id: str, period: str) -> Optional[dict]:
        path = self.segment_ckpt_file(stream_id, period)
        return read_json(path) if os.path.exists(path) else None

    def read_day(self, date: str) -> Optional[dict]:
        path = self.day_file(date)
        return read_json(path) if os.path.exists(path) else None

    def realm(self) -> Dict:
        return read_json(self.realm_file)


def init_realm(layout: Layout, realm_id: str, now_ms: int) -> dict:
    core.check_id(realm_id, "realm_id")
    os.makedirs(layout.root, exist_ok=True)
    os.makedirs(os.path.join(layout.root, "streams"), exist_ok=True)
    os.makedirs(os.path.join(layout.root, "days"), exist_ok=True)
    os.makedirs(layout.anchors_dir(), exist_ok=True)
    if os.path.exists(layout.realm_file):
        existing = layout.realm()
        if existing["realm_id"] != realm_id:
            raise core.FormatError(
                "каталог принадлежит установке %r, а не %r"
                % (existing["realm_id"], realm_id))
        return existing
    obj = {"format": core.FORMAT_VERSION, "hash": core.HASH_NAME,
           "realm_id": realm_id, "created": now_ms}
    write_json(layout.realm_file, obj)
    return obj


def init_stream(layout: Layout, stream_id: str, now_ms: int) -> dict:
    core.check_id(stream_id, "stream_id")
    os.makedirs(layout.segments_dir(stream_id), exist_ok=True)
    path = layout.meta_file(stream_id)
    if os.path.exists(path):
        return read_json(path)
    obj = {"format": core.FORMAT_VERSION, "stream_id": stream_id,
           "genesis": core.genesis(stream_id).hex(), "created": now_ms}
    write_json(path, obj)
    return obj

File: typescript/src/store.ts
/**
 * Раскладка на диске и кодирование записей.
 *
 * Обычные файлы, без базы данных: через семь лет посторонний эксперт должен
 * открыть каталог и понять содержимое без нашей инфраструктуры.
 */

import * as fs from "node:fs";
import * as fsp from "node:fs/promises";
import * as path from "node:path";

import * as core from "./core.js";

export const RECORD_V = 1;

/** Поле записи в том виде, в каком оно лежит в файле. */
export interface StoredField {
  name: string;
  salt?: string;
  s?: string;
  b?: string;
  leaf?: string;
  redacted?: { at: number; seq: number };
}

export interface StoredRecord {
  v: number;
  seq: number;
  ts: number;
  fields: StoredField[];
  content_root: string;
  link: string;
}

export interface SegmentCheckpoint {
  format: string;
  stream_id: string;
  period_id: string;
  first_seq: number;
  last_seq: number;
  count: number;
  link_before: string;
  link_last: string;
  segment_root: string;
  prev_checkpoint: string;
  checkpoint: string;
}

export interface DayCheckpoint {
  format: string;
  realm_id: string;
  date: string;
  streams: string[];
  segment_checkpoints: string[];
  day_root: string;
  prev_checkpoint: string;
  day_checkpoint: string;
}

export type FieldValue = string | Uint8Array;

export function encodeField(name: string, salt: Buffer, value: FieldValue): StoredField {
  const out: StoredField = { name, salt: salt.toString("hex") };
  if (typeof value === "string") out.s = value;
  else if (value instanceof Uint8Array) out.b = Buffer.from(value).toString("base64");
  else throw new core.FormatError(`значение поля ${name} должно быть строкой или байтами`);
  return out;
}

export function rawValue(f: StoredField): Buffer {
  if (f.s !== undefined && f.b !== undefined) {
    throw new core.FormatError(`поле ${f.name}: s и b одновременно`);
  }
  if (f.s !== undefined) return Buffer.from(f.s, "utf8");
  if (f.b !== undefined) return Buffer.from(f.b, "base64");
  throw new core.FormatError(`поле ${f.name}: нет ни s, ни b`);
}

export function fieldsForRoot(fields: StoredField[]): core.LeafInput[] {
  return fields.map((f) => {
    if (f.leaf !== undefined) {
      if (f.salt !== undefined || f.s !== undefined || f.b !== undefined) {
        throw new core.FormatError(
          `вычеркнутое поле ${f.name} сохранило соль или значение`);
      }
      return { name: f.name, leaf: Buffer.from(f.leaf, "hex") };
    }
    if (f.salt === undefined) throw new core.FormatError(`поле ${f.name}: нет соли`);
    return { name: f.name, salt: Buffer.from(f.salt, "hex"), value: rawValue(f) };
  });
}

export function recordLine(rec: StoredRecord): string {
  return JSON.stringify(rec);
}

// --- атомарная запись -------------------------------------------------------

export async function writeJson(file: string, obj: unknown): Promise<void> {
  const tmp = `${file}.tmp`;
  await fsp.mkdir(path.dirname(file), { recursive: true });
  const fh = await fsp.open(tmp, "w");
  try {
    await fh.writeFile(JSON.stringify(obj, null, 2) + "\n", "utf8");
    await fh.sync();
  } finally {
    await fh.close();
  }
  await fsp.rename(tmp, file);
}

export function readJsonSync<T>(file: string): T {
  return JSON.parse(fs.readFileSync(file, "utf8")) as T;
}

export function exists(file: string): boolean {
  return fs.existsSync(file);
}

// --- раскладка --------------------------------------------------------------

export class Layout {
  readonly root: string;

  constructor(root: string) {
    this.root = path.resolve(root);
  }

  get realmFile(): string {
    return path.join(this.root, "realm.json");
  }

  streamDir(id: string): string {
    return path.join(this.root, "streams", id);
  }

  metaFile(id: string): string {
    return path.join(this.streamDir(id), "meta.json");
  }

  segmentsDir(id: string): string {
    return path.join(this.streamDir(id), "segments");
  }

  segmentFile(id: string, period: string): string {
    return path.join(this.segmentsDir(id), `${period}.jsonl`);
  }

  segmentCkptFile(id: string, period: string): string {
    return path.join(this.segmentsDir(id), `${period}.seg.json`);
  }

  dayFile(date: string): string {
    return path.join(this.root, "days", `${date}.day.json`);
  }

  anchorsDir(): string {
    return path.join(this.root, "anchors");
  }

  anchorFile(date: string, kind: string, ext: string): string {
    return path.join(this.anchorsDir(), `${date}.${kind}.${ext}`);
  }

  streams(): string[] {
    const d = path.join(this.root, "streams");
    if (!fs.existsSync(d)) return [];
    return fs.readdirSync(d, { withFileTypes: true })
      .filter((e) => e.isDirectory()).map((e) => e.name).sort();
  }

  periods(id: string): string[] {
    const d = this.segmentsDir(id);
    if (!fs.existsSync(d)) return [];
    return fs.readdirSync(d).filter((n) => n.endsWith(".jsonl"))
      .map((n) => n.slice(0, -6)).sort();
  }

  dates(): string[] {
    const d = path.join(this.root, "days");
    if (!fs.existsSync(d)) return [];
    return fs.readdirSync(d).filter((n) => n.endsWith(".day.json"))
      .map((n) => n.slice(0, -9)).sort();
  }

  anchorsFor(date: string): string[] {
    const d = this.anchorsDir();
    if (!fs.existsSync(d)) return [];
    return fs.readdirSync(d).filter((n) => n.startsWith(`${date}.`))
      .map((n) => path.join(d, n)).sort();
  }

  readSegment(id: string, period: string): StoredRecord[] {
    const file = this.segmentFile(id, period);
    if (!fs.existsSync(file)) return [];
    const out: StoredRecord[] = [];
    const lines = fs.readFileSync(file, "utf8").split("\n");
    lines.forEach((line, i) => {
      if (!line.trim()) return;
      try {
        out.push(JSON.parse(line) as StoredRecord);
      } catch (e) {
        throw new core.FormatError(`${file} строка ${i + 1}: не разбирается как JSON`);
      }
    });
    return out;
  }

  /**
   * Последняя запись отрезка без чтения всего файла.
   *
   * Пишущему нужен только кончик цепи. Читать ради него весь журнал — значит
   * превращать запуск процесса в минуты на журнале за несколько лет. Проверка
   * цепи целиком — работа верификатора, а не писателя.
   */
  readSegmentTail(id: string, period: string): StoredRecord | null {
    const file = this.segmentFile(id, period);
    if (!fs.existsSync(file)) return null;
    const size = fs.statSync(file).size;
    if (size === 0) return null;
    let window = 1 << 16;
    for (;;) {
      const fd = fs.openSync(file, "r");
      const start = Math.max(0, size - window);
      const buf = Buffer.alloc(size - start);
      try {
        fs.readSync(fd, buf, 0, buf.length, start);
      } finally {
        fs.closeSync(fd);
      }
      if (buf[buf.length - 1] !== 0x0a) {
        throw new core.FormatError(
          `${file}: последняя строка оборвана — процесс упал посреди записи. ` +
          "Обрежьте файл до последнего перевода строки; всё до него проверяется " +
          "как обычно.");
      }
      const lines = buf.toString("utf8").split("\n").filter((l) => l.trim());
      if (lines.length) {
        try {
          return JSON.parse(lines[lines.length - 1]!) as StoredRecord;
        } catch {
          // строка обрезана окном — расширяем и пробуем снова
        }
      }
      if (window >= size) {
        throw new core.FormatError(`${file}: последняя запись не разбирается`);
      }
      window = Math.min(window * 8, size);
    }
  }

  readSegmentCkpt(id: string, period: string): SegmentCheckpoint | null {
    const f = this.segmentCkptFile(id, period);
    return fs.existsSync(f) ? readJsonSync<SegmentCheckpoint>(f) : null;
  }

  readDay(date: string): DayCheckpoint | null {
    const f = this.dayFile(date);
    return fs.existsSync(f) ? readJsonSync<DayCheckpoint>(f) : null;
  }

  realm(): { format: string; hash: string; realm_id: string; created: number } {
    return readJsonSync(this.realmFile);
  }
}

export async function initRealm(layout: Layout, realmId: string,
                                now: number): Promise<void> {
  core.checkId(realmId, "realm_id");
  await fsp.mkdir(path.join(layout.root, "streams"), { recursive: true });
  await fsp.mkdir(path.join(layout.root, "days"), { recursive: true });
  await fsp.mkdir(layout.anchorsDir(), { recursive: true });
  if (fs.existsSync(layout.realmFile)) {
    const existing = layout.realm();
    if (existing.realm_id !== realmId) {
      throw new core.FormatError(
        `каталог принадлежит установке ${existing.realm_id}, а не ${realmId}`);
    }
    return;
  }
  await writeJson(layout.realmFile, {
    format: core.FORMAT_VERSION, hash: core.HASH_NAME,
    realm_id: realmId, created: now,
  });
}

export async function initStream(layout: Layout, streamId: string,
                                 now: number): Promise<void> {
  core.checkId(streamId, "stream_id");
  await fsp.mkdir(layout.segmentsDir(streamId), { recursive: true });
  const file = layout.metaFile(streamId);
  if (fs.existsSync(file)) return;
  await writeJson(file, {
    format: core.FORMAT_VERSION, stream_id: streamId,
    genesis: core.genesis(streamId).toString("hex"), created: now,
  });
}

export function periodOf(ms: number): string {
  return new Date(ms).toISOString().slice(0, 10);
}
